File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/IfEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import org.cqframework.cql.elm.visiting.ElmLibraryVisitor;
import org.hl7.elm.r1.If;
import org.opencds.cqf.cql.engine.execution.State;

public class IfEvaluator {
    public static Object internalEvaluate(If elm, State state, ElmLibraryVisitor<Object,State> visitor) {

        Object condition = visitor.visitExpression(elm.getCondition(), state);

        if (condition == null) {
            condition = false;
        }

        return (Boolean) condition ? visitor.visitExpression(elm.getThen(), state) :
                visitor.visitExpression(elm.getElse(), state);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/IfEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import org.cqframework.cql.elm.visiting.ElmLibraryVisitor
import org.hl7.elm.r1.If
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.runtime.Boolean
import org.opencds.cqf.cql.engine.runtime.Value

object IfEvaluator {
    fun internalEvaluate(
        elm: If?,
        state: State?,
        visitor: ElmLibraryVisitor<Value?, State?>,
    ): Value? {
        var condition = visitor.visitExpression(elm!!.condition!!, state) as Boolean?

        if (condition == null) {
            condition = Boolean.FALSE
        }

        return if (condition.value) visitor.visitExpression(elm.then!!, state)
        else visitor.visitExpression(elm.`else`!!, state)
    }
}
